### PR TITLE
Ref [TASK][53322] Get api genres and display screen

### PR DESCRIPTION
**Pull request by ndt0409**
## Related Tickets
- [#Task 53322: Get api genres and display screen](https://edu-redmine.sun-asterisk.vn/issues/53322)
## WHAT
- 
## Evidence (Screenshot or Video)
Screenshot 2022-09-28 110746
## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/WtYPAy |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>  
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>  

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/ndt/themoviedb/data/repository/MovieRepository.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package com.ndt.themoviedb.data.repository
+
+import com.ndt.themoviedb.data.source.MovieDataSource
+import com.ndt.themoviedb.data.source.remote.MovieRemoteDataSource
+import com.ndt.themoviedb.data.source.remote.OnDataLoadedCallback
+import com.ndt.themoviedb.data.source.remote.response.GenresResponse
+import com.ndt.themoviedb.data.source.remote.response.MoviesResponse
+
+class MovieRepository private constructor(
```
Comment (minhnn-1279) [inline app/src/main/java/com/ndt/themoviedb/data/repository/MovieRepository.kt]: không cần private constructor đâu a ơi

File: app/src/main/java/com/ndt/themoviedb/data/source/MovieDataSource.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
+package com.ndt.themoviedb.data.source
+
+import com.ndt.themoviedb.data.source.remote.OnDataLoadedCallback
+import com.ndt.themoviedb.data.source.remote.response.GenresResponse
+import com.ndt.themoviedb.data.source.remote.response.MoviesResponse
+
+interface MovieDataSource {
+
+    interface Local
```
Comment (minhnn-1279) [inline app/src/main/java/com/ndt/themoviedb/data/source/MovieDataSource.kt]: nếu không cần thì remove a nhé

File: app/src/main/java/com/ndt/themoviedb/data/source/remote/response/GenresResponse.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.ndt.themoviedb.data.source.remote.response
+
+import com.ndt.themoviedb.data.model.Genres
+import com.ndt.themoviedb.data.source.remote.fetchjson.ParseDataToObject
+import org.json.JSONObject
+
+data class GenresResponse(var list: List<Genres>) {
```
Comment (minhnn-1279) [inline app/src/main/java/com/ndt/themoviedb/data/source/remote/response/GenresResponse.kt]: cái này để val chứ a ơi :v với a dùng mutable list sau mình clear vs add all data

File: app/src/main/java/com/ndt/themoviedb/data/source/remote/response/MoviesResponse.kt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package com.ndt.themoviedb.data.source.remote.response
+
+import com.ndt.themoviedb.data.model.Movie
+import com.ndt.themoviedb.data.model.MovieResultPage
+import com.ndt.themoviedb.data.source.remote.fetchjson.ParseDataToObject
+import org.json.JSONObject
+
+data class MoviesResponse(var movieResultPage: MovieResultPage, var list: List<Movie>) {
```
Comment (minhnn-1279) [inline app/src/main/java/com/ndt/themoviedb/data/source/remote/response/MoviesResponse.kt]: đây nữa a nhé

File: app/src/main/java/com/ndt/themoviedb/ui/base/BaseAdapter.kt [hunk 5]
```diff
@@ -14,4 +16,26 @@ abstract class BaseAdapter<T, VH : BaseViewHolder<T>> : RecyclerView.Adapter<VH>
 
     private fun getItem(position: Int): T? =
         if (position in 0 until itemCount) items[position] else null
+
+    @SuppressLint("NotifyDataSetChanged")
```
Comment (minhnn-1279) [inline app/src/main/java/com/ndt/themoviedb/ui/base/BaseAdapter.kt]: bỏ suppress lint đi a nhé

File: app/src/main/java/com/ndt/themoviedb/ui/home/HomeFragment.kt [hunk 6]
```diff
@@ -1,10 +1,186 @@
...
+import com.ndt.themoviedb.ui.utils.NetworkUtil
+import com.ndt.themoviedb.ui.utils.OnClickListener
+import com.ndt.themoviedb.ui.utils.UrlConstant
+import kotlinx.android.synthetic.main.fragment_home.view.*
+import java.util.*
+
+class HomeFragment : BaseFragment<FragmentHomeBinding>(FragmentHomeBinding::inflate),
+    HomeContract.View, OnClickListener<Movie> {
+
+    private lateinit var presenter: HomeContract.Presenter
```
Comment (minhnn-1279) [inline app/src/main/java/com/ndt/themoviedb/ui/home/HomeFragment.kt:26]: cái này để by lazy nhé a để chống gán lại presenter

File: gradle.properties [hunk 7]
```diff
@@ -21,5 +21,5 @@ kotlin.code.style=official
 # resources declared in the library itself and none from the library's dependencies,
 # thereby reducing the size of the R class for that library
 android.nonTransitiveRClass=true
-API_KEY=INSERT_YOUR_API_KEY_HERE
+API_KEY=e49c081f7e07e299deda892eeb0c0f75
```
Comment (minhnn-1279) [inline gradle.properties]: https://viblo.asia/p/cac-thu-thuat-de-giau-secret-keypassword-khoi-source-control-va-ung-dung-voi-ci-trong-lap-trinh-android-PwRGgmOnkEd

Chỗ này skip working tree file này a nhé

File: app/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -3,6 +3,9 @@
     xmlns:tools="http://schemas.android.com/tools"
     package="com.ndt.themoviedb">
 
+    <uses-permission android:name="android.permission.INTERNET" />
+    <uses-permission android:name="android.permission.ACCESS_WIFI_STATE" />
+    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
```
Comment (daolq-2712) [inline app/src/main/AndroidManifest.xml:8]:
```suggestion
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    
```

File: app/src/main/java/com/ndt/themoviedb/data/source/remote/fetchjson/ParseDataToObject.kt [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+import com.ndt.themoviedb.data.model.MovieResultPage
+import org.json.JSONObject
+
+object ParseDataToObject {
+    fun parJsonToMovieResultPage(jsonObject: JSONObject): MovieResultPage {
+        return MovieResultPage(jsonObject)
+    }
+
+    fun parJsonToMovies(jsonObject: JSONObject): MutableList<Movie> {
+        val list = mutableListOf<Movie>()
```
Comment (daolq-2712) [inline app/src/main/java/com/ndt/themoviedb/data/source/remote/fetchjson/ParseDataToObject.kt]:
```suggestion
        val movies = mutableListOf<Movie>()
```

File: app/src/main/java/com/ndt/themoviedb/data/source/remote/fetchjson/ParseDataToObject.kt [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+        val jsonArray = jsonObject.getJSONArray(Movie.MovieEntry.MOVIE)
+        for (i in 0 until jsonArray.length()) {
+            list.add(Movie(jsonArray.getJSONObject(i)))
+        }
+        return list
+    }
+
+    fun parJsonToGenres(jsonObject: JSONObject): MutableList<Genres> {
+        val list = mutableListOf<Genres>()
```
Comment (daolq-2712) [inline app/src/main/java/com/ndt/themoviedb/data/source/remote/fetchjson/ParseDataToObject.kt]:
```suggestion
        val listGene = mutableListOf<Genres>()
```

File: app/src/main/java/com/ndt/themoviedb/data/source/remote/response/MoviesResponse.kt [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package com.ndt.themoviedb.data.source.remote.response
+
+import com.ndt.themoviedb.data.model.Movie
+import com.ndt.themoviedb.data.model.MovieResultPage
+import com.ndt.themoviedb.data.source.remote.fetchjson.ParseDataToObject
+import org.json.JSONObject
+
+data class MoviesResponse(val movieResultPage: MovieResultPage, val mutableList: MutableList<Movie>) {
```
Comment (daolq-2712) [inline app/src/main/java/com/ndt/themoviedb/data/source/remote/response/MoviesResponse.kt]:
```suggestion
data class MoviesResponse(val movieResultPage: MovieResultPage, val listMovie: MutableList<Movie>) {
```

File: app/src/main/java/com/ndt/themoviedb/data/source/remote/response/GenresResponse.kt [hunk 12]
```diff
@@ -0,0 +1,11 @@
+package com.ndt.themoviedb.data.source.remote.response
+
+import com.ndt.themoviedb.data.model.Genres
+import com.ndt.themoviedb.data.source.remote.fetchjson.ParseDataToObject
+import org.json.JSONObject
+
+data class GenresResponse(val mutableList: MutableList<Genres>) {
```
Comment (daolq-2712) [inline app/src/main/java/com/ndt/themoviedb/data/source/remote/response/GenresResponse.kt]:
```suggestion
data class GenresResponse(val listGene: MutableList<Genres>) {
```

File: app/src/main/java/com/ndt/themoviedb/ui/home/HomeFragment.kt [hunk 13]
```diff
@@ -1,10 +1,184 @@
...
+    private fun initAdapter() {
+        view?.run {
+            viewBinding.sliderViewPager.adapter = movieSlideAdapter
+            indicatorTabLayout.setupWithViewPager(
+                viewBinding.sliderViewPager,
+                true
+            )
+        }
+        movieSlideAdapter.setSlideItemClickListener(this)
+        Timer().scheduleAtFixedRate(object : TimerTask() {
```
Comment (daolq-2712) [inline app/src/main/java/com/ndt/themoviedb/ui/home/HomeFragment.kt:138]: xử lý leak memory

File: app/src/main/java/com/ndt/themoviedb/ui/home/HomePresenter.kt [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+import com.ndt.themoviedb.data.source.remote.response.GenresResponse
+import com.ndt.themoviedb.data.source.remote.response.MoviesResponse
+import com.ndt.themoviedb.ui.utils.UrlConstant
+
+class HomePresenter(private val movieRepository: MovieRepository) : HomeContract.Presenter {
+    private var view: HomeContract.View? = null
+
+    override fun onStart() {
+        view?.onLoading(false)
+        getGenres()
```
Comment (daolq-2712) [inline app/src/main/java/com/ndt/themoviedb/ui/home/HomePresenter.kt:14]: gọi ở onStart của view

File: app/src/main/java/com/ndt/themoviedb/ui/home/adapter/SliderViewPagerAdapter.kt [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+import com.ndt.themoviedb.data.model.Movie
+import com.ndt.themoviedb.ui.home.HomeFragment
+import com.ndt.themoviedb.ui.utils.GetImageAsyncTask
+import com.ndt.themoviedb.ui.utils.OnClickListener
+import com.ndt.themoviedb.ui.utils.OnFetchImageListener
+import com.ndt.themoviedb.ui.utils.UrlConstant
+import kotlinx.android.synthetic.main.item_slide.view.*
+
+class SliderViewPagerAdapter : PagerAdapter() {
+    private val list = arrayListOf<Movie>()
```
Comment (daolq-2712) [inline app/src/main/java/com/ndt/themoviedb/ui/home/adapter/SliderViewPagerAdapter.kt]: khai báo biến tường minh

File: app/src/main/java/com/ndt/themoviedb/ui/home/adapter/SliderViewPagerAdapter.kt [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+        slideLayout.titleSlideTextView.text = list[position].title
+        container.addView(slideLayout)
+        return slideLayout
+    }
+
+    private fun getImageSlide(
+        img: ImageView,
+        movie: Movie?
+    ) {
+        GetImageAsyncTask(
```
Comment (daolq-2712) [inline app/src/main/java/com/ndt/themoviedb/ui/home/adapter/SliderViewPagerAdapter.kt:42]: xử lý leak memory

**Review by daolq-2712 — CHANGES_REQUESTED** [review]
Pls check my comments!

**Review by minhnn-1279 — APPROVED** [review]

**Pull request merged**